metricshub-ping-extension/src/main/java/org/sentrysoftware/metricshub/extension/ping/PingExtension.java

```diff
@@ -0,0 +1,191 @@
+package org.sentrysoftware.metricshub.extension.ping;
+
+/*-
+ * ╱╲╱╲╱╲╱╲╱╲╱╲╱╲╱╲╱╲╱╲╱╲╱╲╱╲╱╲╱╲╱╲╱╲╱╲╱╲╱╲
+ * MetricsHub Ping Extension
+ * ჻჻჻჻჻჻
+ * Copyright 2023 - 2024 Sentry Software
+ * ჻჻჻჻჻჻
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ * ╲╱╲╱╲╱╲╱╲╱╲╱╲╱╲╱╲╱╲╱╲╱╲╱╲╱╲╱╲╱╲╱╲╱╲╱╲╱╲╱
+ */
+
+import com.fasterxml.jackson.databind.DeserializationFeature;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.MapperFeature;
+import com.fasterxml.jackson.databind.SerializationFeature;
+import com.fasterxml.jackson.databind.json.JsonMapper;
+import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.UnaryOperator;
+import lombok.extern.slf4j.Slf4j;
+import org.sentrysoftware.metricshub.engine.common.exception.InvalidConfigurationException;
+import org.sentrysoftware.metricshub.engine.configuration.IConfiguration;
+import org.sentrysoftware.metricshub.engine.connector.model.identity.criterion.Criterion;
+import org.sentrysoftware.metricshub.engine.connector.model.monitor.task.source.Source;
+import org.sentrysoftware.metricshub.engine.extension.IProtocolExtension;
+import org.sentrysoftware.metricshub.engine.strategy.detection.CriterionTestResult;
+import org.sentrysoftware.metricshub.engine.strategy.source.SourceTable;
+import org.sentrysoftware.metricshub.engine.telemetry.MetricFactory;
+import org.sentrysoftware.metricshub.engine.telemetry.Monitor;
+import org.sentrysoftware.metricshub.engine.telemetry.TelemetryManager;
+
+/**
+ * This class implements the {@link IProtocolExtension} contract, reports the supported features,
+ * checks the reachability of a host through a ICMP Ping.
+ */
+@Slf4j
+public class PingExtension implements IProtocolExtension {
+
+	private PingRequestExecutor pingRequestExecutor;
+
+	public PingExtension() {
+		pingRequestExecutor = new PingRequestExecutor();
+	}
+
+	/**
+	 * Protocol up status value '1.0'
+	 */
+	public static final Double UP = 1.0;
+
+	/**
+	 * Protocol down status value '0.0'
+	 */
+	public static final Double DOWN = 0.0;
+
+	/**
+	 * ICMP Ping Up metric name format that will be saved by the metric factory
+	 */
+	public static final String PING_UP_METRIC = "metricshub.host.up{protocol=\"ping\"}";

```

Comment:
I would expect this to be decided by the agent, not in the protocol extension.

Reply:
No need to make this metric name configurable.

But I was thinking that the metric name should not be decided by the protocol itself, but by the agent, where the code would look like:

```java
foreach (Protocol protocol : config.getProtocols()) {
    boolean isAvailable = protocol.checkProtocol();
    telemetryManager.addPoint("metricshub.host.protocol.up{protocol=\" + protocol.getName() + "\"}", isAvailable ? 1 : 0);
}
```

This will avoid passing a TelemetryManager instance in each protocol extension, as well as all of the OpenTelemetry-specific code.

Reply:
I understand now 😉! 

The OpenTelemetry-specific code is not passed to the protocol extension. Currently, the OpenTelemetry code observes the metric's value in the agent before sending it to the OTLP receiver.

The code you suggested is already implemented in the agent through the `ProtocolHealthCheckStrategy`, which is triggered at each collect cycle, except for the `metricshub.host.protocol.up` collect.

```java
// Call the extensions to check the protocol health
final List<IProtocolExtension> protocolExtensions = extensionManager.findProtocolCheckExtensions(telemetryManager);
protocolExtensions.forEach(protocolExtension -> protocolExtension.checkProtocol(telemetryManager));
}
```

Indeed, the `ProtocolHealthCheckStrategy` can be updated to collect the protocol point, this will avoid code duplication in each extension. Several pieces of information are required from the TelemetryManager: hostname, configuration, so this can be passed as method arguments.

```java
// Call the extensions to check the protocol health
final List<IProtocolExtension> protocolExtensions = extensionManager.findProtocolCheckExtensions(hostname, configuration);
protocolExtensions.forEach(protocolExtension -> { 

    final Monitor hostMonitor = telemetryManager.getEndpointHostMonitor();
    final int state = protocolExtension.checkProtocol(telemetryManager);
    new MetricFactory()
        .collectNumberMetric(
            hostMonitor,
            "metricshub.host.protocol.up{protocol=\" + protocolExtension.getIdentifier() + \"}",
            state,
            telemetryManager.getStrategyTime()
        );
});
}
```

This can be done in a separate issue as it is a refactoring code.